### PR TITLE
(plugin) os::linux::snmp - add service discovery listdiskio

### PR DESCRIPTION
**Pull request by sdepassio**
## Description

Add a Service Discovery Rule to discover Disk IO Device 

Mon-17792

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

<h2> How this pull request can be tested ? </h2>

Tests in the os-linux-snmp.robot file.

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Comment by garnier-quentin** [comment]
There are too many files ;)

File: .github/workflows/tests-functional.yml [hunk 1]
```diff
@@ -12,58 +12,65 @@ on:
...
   functional-tests-with-robot:
     runs-on: ubuntu-22.04
 
     steps:
       - uses: actions/checkout@v3
+
+      - name: Install libs
+        run: |
+          sudo apt update
+          sudo apt upgrade -y
```
Comment (kduret) [inline .github/workflows/tests-functional.yml:27]: do not upgrade runners

File: tests/functional/snmp/os-linux-snmp.robot [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            ...    ${output}
+            ...    label
+            Should Be Equal As Integers
+            ...    ${list_diskio_test.nbresults}
+            ...    ${nb_results}
+            ...    msg=Wrong output result for list diskio devices: ${list_diskio_test}.{\n}Command output:{\n}${output}
+        END
+        ${length} =    Get Length    ${list_diskio_test.result_match}
+        IF    ${length} > 0
+            @{labels} =    Get Elements    ${output}    label
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:178]:
```suggestion
            @{labels}    Get Elements    ${output}    label
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+        ${output} =    Run    ${command}
+        ${length} =    Get Length    ${list_diskio_test.nbresults}
+        IF    ${length} > 0
+            ${nb_results} =    Get Element Count
+            ...    ${output}
+            ...    label
+            Should Be Equal As Integers
+            ...    ${list_diskio_test.nbresults}
+            ...    ${nb_results}
+            ...    msg=Wrong output result for list diskio devices: ${list_diskio_test}.{\n}Command output:{\n}${output}
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:174]:
```suggestion
            ...    Wrong output result for list diskio devices: ${list_diskio_test}.{\n}Command output:{\n}${output}
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+            ${command} =    Catenate    ${command}    --display-transform-src=${list_diskio_test.display_transform_src}
+        END
+        ${length} =    Get Length    ${list_diskio_test.display_transform_dst}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --display-transform-dst=${list_diskio_test.display_transform_dst}
+        END
+        ${output} =    Run    ${command}
+        ${length} =    Get Length    ${list_diskio_test.nbresults}
+        IF    ${length} > 0
+            ${nb_results} =    Get Element Count
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:168]:
```suggestion
            ${nb_results}    Get Element Count
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        END
+        ${length} =    Get Length    ${list_diskio_test.regexp_isensitive}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --regexp-isensitive
+        END
+        ${length} =    Get Length    ${list_diskio_test.display_transform_src}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --display-transform-src=${list_diskio_test.display_transform_src}
+        END
+        ${length} =    Get Length    ${list_diskio_test.display_transform_dst}
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:161]:
```suggestion
        ${length}    Get Length    ${list_diskio_test.display_transform_dst}
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --regexp
+        END
+        ${length} =    Get Length    ${list_diskio_test.regexp_isensitive}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --regexp-isensitive
+        END
+        ${length} =    Get Length    ${list_diskio_test.display_transform_src}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --display-transform-src=${list_diskio_test.display_transform_src}
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:159]:
```suggestion
            ${command}    Catenate    ${command}    --display-transform-src=${list_diskio_test.display_transform_src}
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        END
+        ${length} =    Get Length    ${list_diskio_test.regexp}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --regexp
+        END
+        ${length} =    Get Length    ${list_diskio_test.regexp_isensitive}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --regexp-isensitive
+        END
+        ${length} =    Get Length    ${list_diskio_test.display_transform_src}
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:157]:
```suggestion
        ${length}    Get Length    ${list_diskio_test.display_transform_src}
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --name
+        END
+        ${length} =    Get Length    ${list_diskio_test.regexp}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --regexp
+        END
+        ${length} =    Get Length    ${list_diskio_test.regexp_isensitive}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --regexp-isensitive
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:155]:
```suggestion
            ${command}    Catenate    ${command}    --regexp-isensitive
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+        END
+        ${length} =    Get Length    ${list_diskio_test.name}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --name
+        END
+        ${length} =    Get Length    ${list_diskio_test.regexp}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --regexp
+        END
+        ${length} =    Get Length    ${list_diskio_test.regexp_isensitive}
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:153]:
```suggestion
        ${length}    Get Length    ${list_diskio_test.regexp_isensitive}
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --diskiodevice=${list_diskio_test.diskiodevice}
+        END
+        ${length} =    Get Length    ${list_diskio_test.name}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --name
+        END
+        ${length} =    Get Length    ${list_diskio_test.regexp}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --regexp
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:151]:
```suggestion
            ${command}    Catenate    ${command}    --regexp
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        ...    --snmp-community=os_linux_snmp_plugin
+        ...    --snmp-port=2024
+        ...    --disco-show
+        ${length} =    Get Length    ${list_diskio_test.diskiodevice}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --diskiodevice=${list_diskio_test.diskiodevice}
+        END
+        ${length} =    Get Length    ${list_diskio_test.name}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --name
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:147]:
```suggestion
            ${command}    Catenate    ${command}    --name
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 12]
```diff
@@ -0,0 +1,183 @@
...
+        ...    --hostname=127.0.0.1
+        ...    --snmp-version=2
+        ...    --snmp-community=os_linux_snmp_plugin
+        ...    --snmp-port=2024
+        ...    --disco-show
+        ${length} =    Get Length    ${list_diskio_test.diskiodevice}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --diskiodevice=${list_diskio_test.diskiodevice}
+        END
+        ${length} =    Get Length    ${list_diskio_test.name}
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:145]:
```suggestion
        ${length}    Get Length    ${list_diskio_test.name}
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 13]
```diff
@@ -0,0 +1,183 @@
...
+        ...    ${CMD}
+        ...    --mode=list-diskio
+        ...    --hostname=127.0.0.1
+        ...    --snmp-version=2
+        ...    --snmp-community=os_linux_snmp_plugin
+        ...    --snmp-port=2024
+        ...    --disco-show
+        ${length} =    Get Length    ${list_diskio_test.diskiodevice}
+        IF    ${length} > 0
+            ${command} =    Catenate    ${command}    --diskiodevice=${list_diskio_test.diskiodevice}
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:143]:
```suggestion
            ${command}    Catenate    ${command}    --diskiodevice=${list_diskio_test.diskiodevice}
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 14]
```diff
@@ -0,0 +1,183 @@
...
+    FOR    ${list_diskio_test}    IN    @{list_diskio_tests}
+        ${command} =    Catenate
+        ...    ${CMD}
+        ...    --mode=list-diskio
+        ...    --hostname=127.0.0.1
+        ...    --snmp-version=2
+        ...    --snmp-community=os_linux_snmp_plugin
+        ...    --snmp-port=2024
+        ...    --disco-show
+        ${length} =    Get Length    ${list_diskio_test.diskiodevice}
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:141]:
```suggestion
        ${length}    Get Length    ${list_diskio_test.diskiodevice}
```

**Review by kduret — APPROVED** [review]
LGTM for workflow changes

File: tests/functional/snmp/os-linux-snmp.robot [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+...                         &{list_diskio_test10}
+...                         &{list_diskio_test11}
+
+
+*** Test Cases ***
+Linux SNMP list diskio devices
+    [Documentation]    List Linux diskio devices
+    [Tags]    os    linux    snmp
+    FOR    ${list_diskio_test}    IN    @{list_diskio_tests}
+        ${command} =    Catenate
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:133]:
```suggestion
        ${command}    Catenate
```

File: tests/functional/snmp/os-linux-snmp.robot [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+        IF    ${length} > 0
+            ${nb_results} =    Get Element Count
+            ...    ${output}
+            ...    label
+            Should Be Equal As Integers
+            ...    ${list_diskio_test.nbresults}
+            ...    ${nb_results}
+            ...    msg=Wrong output result for list diskio devices: ${list_diskio_test}.{\n}Command output:{\n}${output}
+        END
+        ${length} =    Get Length    ${list_diskio_test.result_match}
```
Comment (thearchit3ct) [inline tests/functional/snmp/os-linux-snmp.robot:176]:
```suggestion
        ${length}    Get Length    ${list_diskio_test.result_match}
```

**Review by thearchit3ct — APPROVED** [review]

**Review by thearchit3ct — APPROVED** [review]

File: src/snmp_standard/mode/listdiskio.pm [hunk 17]
```diff
@@ -0,0 +1,188 @@
...
+    bless $self, $class;
+
+    $options{options}->add_options(arguments => {
+        'diskiodevice:s'          => { name => 'diskiodevice' },
+        'name'                    => { name => 'use_name' },
+        'regexp'                  => { name => 'use_regexp' },
+        'regexp-isensitive'       => { name => 'use_regexpi' },
+        'display-transform-src:s' => { name => 'display_transform_src' },
+        'display-transform-dst:s' => { name => 'display_transform_dst' }
+    });
```
Comment (garnier-quentin) [inline src/snmp_standard/mode/listdiskio.pm:40]: In discovery, we dont need all that options. The filter is done in the discovery module if needed.

**Review by garnier-quentin — COMMENTED** [review]
Why there are some changes on cloud files for that PR also ?

**Review by lucie-dubrunfaut — DISMISSED** [review]

**Comment by sdepassio** [comment]
> Why there are some changes on cloud files for that PR also ?

To add SNMP tests the other test files have been moved and a small improvement requested on this PR was reported on all tests files.

**Review by lucie-dubrunfaut — DISMISSED** [review]

**Review by garnier-quentin — APPROVED** [review]

**Review by lucie-dubrunfaut — APPROVED** [review]

**Pull request merged**